apps/shared/avifjpeg.c

```diff
@@ -25,6 +28,176 @@ static void my_error_exit(j_common_ptr cinfo)
+
+    // lines of output image attempt to read per jpeg_read_raw_data call
+    int readLines = 0;
+    // lines of sample to read per call for each channel
+    int linesPerCall[3] = { 0, 0, 0 };
+    // lines of sample expect to get for each channel
+    int targetRead[3] = { 0, 0, 0 };
+    // already read lines for each channel
+    int alreadyRead[3] = { 0, 0, 0 };
+    // which avif channel to write to for each jpeg channel
+    enum avifChannelIndex targetChannel[3] = { AVIF_CHAN_R, AVIF_CHAN_R, AVIF_CHAN_R };
+    for (int i = 0; i < cinfo->num_components; ++i) {
+        jpeg_component_info * comp = &cinfo->comp_info[i];
+
+        linesPerCall[i] = comp->v_samp_factor * comp->AVIF_LIBJPEG_DCT_v_scaled_size;
+        targetRead[i] = comp->downsampled_height;
+        buffer[i] = (*cinfo->mem->alloc_sarray)((j_common_ptr)cinfo,
+                                                JPOOL_IMAGE,
+                                                comp->width_in_blocks * comp->AVIF_LIBJPEG_DCT_h_scaled_size,
+                                                linesPerCall[i]);
+        readLines = AVIF_MAX(readLines, linesPerCall[i]);
+    }
+
+    avifImageAllocatePlanes(avif, AVIF_PLANES_YUV);
+
+    if (cinfo->jpeg_color_space == JCS_YCbCr) {
+        targetChannel[0] = AVIF_CHAN_Y;
+        targetChannel[1] = AVIF_CHAN_U;
+        targetChannel[2] = AVIF_CHAN_V;
+    } else if (cinfo->jpeg_color_space == JCS_GRAYSCALE) {
+        targetChannel[0] = AVIF_CHAN_Y;
+    } else /* cinfo->jpeg_color_space == JCS_RGB */ {
+        targetChannel[0] = AVIF_CHAN_V;
+        targetChannel[1] = AVIF_CHAN_Y;
+        targetChannel[2] = AVIF_CHAN_U;
+    }
+
+    int workComponents = avif->yuvFormat == AVIF_PIXEL_FORMAT_YUV400 ? 1 : cinfo->num_components;
+
+    while (cinfo->output_scanline < cinfo->output_height) {
+        jpeg_read_raw_data(cinfo, buffer, readLines);
+
+        for (int i = 0; i < workComponents; ++i) {
+            int linesRead = AVIF_MIN(targetRead[i] - alreadyRead[i], linesPerCall[i]);
+            for (int j = 0; j < linesRead; ++j) {
+                memcpy(&avif->yuvPlanes[targetChannel[i]][avif->yuvRowBytes[targetChannel[i]] * (alreadyRead[i] + j)],
+                       buffer[i][j],
+                       avif->yuvRowBytes[targetChannel[i]]);
+            }
+            alreadyRead[i] += linesPerCall[i];
+        }
+    }
+
+    jpeg_finish_decompress(cinfo);
+}
+
+static avifBool avifJPEGReadCopy(avifImage * avif, struct jpeg_decompress_struct * cinfo)
+{
+    if (avif->depth != 8) {
+        return AVIF_FALSE;
+    }
+
+    if (cinfo->jpeg_color_space == JCS_YCbCr) {
+        // Import from YUV: must using compatible matrixCoefficients.
+        if ((avif->matrixCoefficients == AVIF_MATRIX_COEFFICIENTS_BT601) ||
+            (avif->matrixCoefficients == AVIF_MATRIX_COEFFICIENTS_CHROMA_DERIVED_NCL &&
+             avif->colorPrimaries == AVIF_COLOR_PRIMARIES_BT470M)) {
```

Comment:
I think checking against `MC=12` here might not be worth it, and I don't believe `AVIF_COLOR_PRIMARIES_BT470M` is correct either way? The derived matrix coefficients for BT601 either come from `AVIF_COLOR_PRIMARIES_BT470BG` or `AVIF_COLOR_PRIMARIES_BT601` (I'd have to do the math to remember precisely), but I'm almost positive it isn't `AVIF_COLOR_PRIMARIES_BT470M`.

Reply:
Alright, I did the math, and you're right:

    calcYUVCoefficients = (primaries) ->
      rX = primaries[0]
      rY = primaries[1]
      gX = primaries[2]
      gY = primaries[3]
      bX = primaries[4]
      bY = primaries[5]
      wX = primaries[6]
      wY = primaries[7]
      rZ = 1.0 - (rX + rY) # (Eq. 34)
      gZ = 1.0 - (gX + gY) # (Eq. 35)
      bZ = 1.0 - (bX + bY) # (Eq. 36)
      wZ = 1.0 - (wX + wY) # (Eq. 37)
      kr = (rY * (wX * (gY * bZ - bY * gZ) + wY * (bX * gZ - gX * bZ) + wZ * (gX * bY - bX * gY))) /
           (wY * (rX * (gY * bZ - bY * gZ) + gX * (bY * rZ - rY * bZ) + bX * (rY * gZ - gY * rZ)))

      # (Eq. 32)
      kb = (bY * (wX * (rY * gZ - gY * rZ) + wY * (gX * rZ - rX * gZ) + wZ * (rX * gY - gX * rY))) /
           (wY * (rX * (gY * bZ - bY * gZ) + gX * (bY * rZ - rY * bZ) + bX * (rY * gZ - gY * rZ)))

      # (Eq. 33)
      coeffs = []
      coeffs[0] = kr
      coeffs[2] = kb
      coeffs[1] = 1.0 - coeffs[0] - coeffs[2]

      return coeffs

    main = ->
      primaries =
        "BT470M":  [ 0.67, 0.33, 0.21, 0.71, 0.14, 0.08, 0.310, 0.316 ]
        "BT470BG": [ 0.64, 0.33, 0.29, 0.60, 0.15, 0.06, 0.3127, 0.3290 ]
        "BT601":   [ 0.630, 0.340, 0.310, 0.595, 0.155, 0.070, 0.3127, 0.3290 ]

      console.log "Looking for 0.299, 0.114:\n"

      for name, p of primaries
        c = calcYUVCoefficients(p)
        console.log "#{name}: Kr: #{c[0].toFixed(4)} Kb: #{c[2].toFixed(4)}"

    main()

... outputs:

    Looking for 0.299, 0.114:

    BT470M: Kr: 0.2990 Kb: 0.1146
    BT470BG: Kr: 0.2220 Kb: 0.0713
    BT601: Kr: 0.2124 Kb: 0.0866

So you're right that `AVIF_COLOR_PRIMARIES_BT470M` happens to match. That said, I don't think CICP `4/x/12` is an interesting combination to check. Perhaps if we come up with a solution for the "enduser expectations" issue cited above, this might end up being fine.

Reply:
> The derived matrix coefficients for BT601 either come from

You know how the matrix coeff. are derived from primaries? Can you paste on Wikipedia? I set two citation needed in YCbCr article for that. https://en.wikipedia.org/wiki/YCbCr#ITU-R_BT.709_conversion

Now, from T.871 you can read that JPEG uses full range JPEG that uses BT.601 matrix (coeff. of 0.299, 0.587, 0.114) and after that without ICC profile sRGB is assumed. That means not 2.2, but almost 2.2 piesewise with linear segment function and BT.709 primaries.

It is NOT 470M since 470M uses (from ITU-T H.273) perfect 2.2 gamma and aboslutely different primaries, including DIFFERENT WHITE POINT, white C and not even series D of illuminats.

Reply:
I've supplied the math I did for my check afterwards. I think we should avoid offering this functionality when `MC=12`, to be safe.

Reply:
Oh, that is page 10 of ITU-T H.273! Sigh.

Reply:
Wait a second, is not those equestions only for "Otherwise (MatrixCoefficients is equal to 12 or 13)," which are for very special "Chromaticity-derived non-constant luminance system"? Or are those others also derived like that?

Reply:
(I'm not sure if this was a question for me or not, but I will supply some info and hope I answer something for you.)

For any `AVIF_MATRIX_COEFFICIENTS_*` enum that ends up being a typical matrix multiply, we have a table of coefficients that simply cite the values in Table 4 of H.273:

https://github.com/AOMediaCodec/libavif/blob/master/src/colr.c#L81

libavif won't do any of that math on MC=1, for example, it'll just use whatever is in that table. That said, you can certainly start with any set of color primaries and (using the math above or in libavif's `calcYUVInfoFromCICP()`), you can derive the coefficients. They correspond exactly for CP/MC=1 (BT709 and CP/MC=9 (BT2020), for example.

Reply:
Yeah, I already found the originated source with proper equestions. The derivation of the matrix from primaries is called normalized primary matrix and then luminance equation for Y' is derived. Yeah. It is SMPTE RP 177.

Just a little bit of linaer algebra.

Reply:
> I've supplied the math I did for my check afterwards. I think we should avoid offering this functionality when `MC=12`, to be safe.

My consideration is, if YUV->RGB->YUV gives the same result, we just skip it and use YUV directly.

Reply:
"BT601":   [ 0.630, 0.340, 0.310, 0.595, 0.155, 0.070, 0.3127, 0.3290 ]

>these are actually BT.709 primaries...

Reply:
No, they aren't.

From H.273:

![image](https://user-images.githubusercontent.com/332342/114254861-dab5e600-9966-11eb-8492-32b9a60728f6.png)

I appreciate your interest in this particular PR and curiosities surrounding color math, but this is way out of scope, and incorrect. Please open a fresh GitHub issue if you discover incorrect constants in libavif.

Reply:
Yeah, you right. It is for SMPTE 270M matrix. Wow.

Reply:
Now, I do not understand how DCI-D65 primaries that are used in Netflix are used with BT.2020-ncl matrix. Should not a different matrix be used? Uuurgh.

Reply:
Wow, try to google: 

BT470BG: Kr: 0.2220 Kb: 0.0713
Apparently it was defined in 1987, even though is not known now! Wow. Haha.

Also BT.709-1 (that is version 1) used slighly different coeff. from version 2, interesting why. (See wiki again.)